pdf/pdfselection.cpp

```diff
@@ -0,0 +1,486 @@
+ * Contact: Damien Caliste <dcaliste@free.fr>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2 only.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include "pdfselection.h"
+#include "pdfdocument.h"
+
+class PDFSelection::Private
+{
+public:
+    Private()
+        : canvas(nullptr)
+        , pageIndexStart(-1)
+        , boxIndexStart(-1)
+        , pageIndexStop(-1)
+        , boxIndexStop(-1)
+        , wiggle(4.)
+    {
+    }
+
+    PDFCanvas *canvas;
+
+    QPointF start;
+    QPointF stop;
+
+    int pageIndexStart;
+    int boxIndexStart;
+
+    int pageIndexStop;
+    int boxIndexStop;
+
+    bool handleReversed;
+
+    qreal wiggle;
+
+    enum Position {
+        At,
+        Before,
+        After
+    };
+    void textBoxId(const QPointF &point, Position position, int *pageIndex, int *boxIndex);
+    int sliceCount(int pageIndex1, int boxIndex1, int pageIndex2, int boxIndex2);
+};
+
+PDFSelection::PDFSelection(QObject *parent)
+  : QAbstractListModel(parent), d(new Private)
+{
+}
+
+PDFSelection::~PDFSelection()
+{
+    delete d;
+}
+
+QHash<int, QByteArray> PDFSelection::roleNames() const
+{
+    QHash<int, QByteArray> roles;
+
+    roles.insert(Rect, "rect");
+    roles.insert(Text, "text");
+
+    return roles;
+}
+
+QVariant PDFSelection::data(const QModelIndex& index, int role) const
+{
+    PDFDocument *doc = d->canvas->document();
+    int pageIndex = -1;
+    int boxIndex;
+    QVariant result;
+
+    if (index.isValid()) {
+        boxIndex = index.row();
+        if (boxIndex > -1 && boxIndex < count()) {
+            boxIndex += d->boxIndexStart;
+            for (pageIndex = d->pageIndexStart;
+                 pageIndex <= d->pageIndexStop; pageIndex++) {
+                int nBoxes = doc->textBoxesAtPage(pageIndex).count();
+                if (boxIndex < nBoxes) {
+                    break;
+                } else {
+                    boxIndex -= nBoxes;
+                }
+            }
+        }
+    }
+    if (pageIndex >= 0) {
+        const PDFDocument::TextList &boxes = doc->textBoxesAtPage(pageIndex);
+        const QPair<QRectF, Poppler::TextBox*> &box = boxes.at(boxIndex);
+        switch(role) {
+        case Rect:
+            result.setValue<QRectF>(d->canvas->fromPageToItem(pageIndex, box.first));
+            break;
+        case Text:
+            result.setValue<QString>(box.second->text());
+            break;
```

Comment:
Nah, doesn't matter much being there or not, only few lines difference. However you like.